### DISCUSSION TITLE
what is the diffrence  between sycl::context and pi_context?

### DISCUSSION THREAD

**Discussion by wangzy0327** [intel/llvm#7328] [Q&A]
what is the diffrence  between sycl::context and pi_context?

**Comment by AlexeySachkov** [comment]
Hi @wangzy0327, `sycl::context` is a class which is documented by the [SYCL core](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:interface.context.class) and it should be used when you develop your SYCL/DPC++ applications.

`pi_context` is an internal opaque type which our implementation of SYCL RT uses. Our runtime is built on top of a portability layer which we call PI: Plugin Interface. It allows us to have a unified codebase at SYCL RT level, but use different low-level runtimes (like OpenCL, Level Zero, CUDA, etc.) to implement SYCL APIs.

So, SYCL RT provides an implementation of `sycl::context`, which uses `pi_context` under the hood as a context identifier and that `pi_context` is treated differently depending on which low-level runtime (or PI plugin, or SYCL backend) is being used.
